### ISSUE TITLE
Can't build Testbed on Windows using CMake; missing PKG_CONFIG_EXECUTABLE

### ISSUE THREAD

**Issue by NauticalMile64** [louis-langholtz/PlayRho#16]
### Expected/Desired Behavior or Experience:
I would like to be able to build using CMake on Windows so I can address the unit test linking issue #1.

### Actual Behavior:
CMake returns with the following errors (after specifying the glew library location correctly):

```
PLAYRHO_REALNUM_TYPE=float
CMake Error at C:/Program Files (x86)/CMake/share/cmake-3.9/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
Call Stack (most recent call first):
  C:/Program Files (x86)/CMake/share/cmake-3.9/Modules/FindPackageHandleStandardArgs.cmake:377 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files (x86)/CMake/share/cmake-3.9/Modules/FindPkgConfig.cmake:36 (find_package_handle_standard_args)
  Testbed/CMakeLists.txt:4 (find_package)
```

### Steps to Reproduce the Actual Behavior:
I found this behavior while following the old CMake setup instructions. I've installed pkg-config using information found in the answers to [this question](https://stackoverflow.com/questions/1710922/how-to-install-pkg-config-in-windows) (along with the necessary `.dll` files, verifying that the program runs), and adding `PKG_CONFIG_EXECUTABLE=C:\path\to\pkg-config.exe` in the CMake GUI. When I configure, it deletes the entry and reports the same error as above.

**Comment by NauticalMile64** [comment]
I have CMake configuring for VS 2017 x64

**Comment by louis-langholtz** [comment]
Does this problem occur when using CMake to build just the PlayRho library, or when you're using CMake to also build the Testbed?

**Comment by NauticalMile64** [comment]
You're right, it only occurs when I try configure to build the testbed. When I select everything except the testbed, it generates the VS solution with only the `PLAYRHO_REALNUM_TYPE=float` message displayed in red.

Also, you (or CMake, I'm not sure which) generate four different build configurations. Is this intended? Should I try and get everything working for each configuration? (There are build errors I will submit as separate issues)

**Comment by louis-langholtz** [comment]
@NauticalMile64 Would you be able to attach any kind of screen dump of the output that you're describing (like a snap shot of the relevant window)? I think that might be very helpful for me and anyone else who runs into anything similar and we could probably re-use those in documentation/instructions.

**Comment by louis-langholtz** [comment]
I think the output in red that was mentioned is from [line 4 of the `PlayRho/PlayRho/CMakeLists.txt` file](https://github.com/louis-langholtz/PlayRho/blob/master/PlayRho/CMakeLists.txt#L4) which states:

    message("PLAYRHO_REALNUM_TYPE=${PLAYRHO_REALNUM_TYPE}")

I had added that to be an output diagnostic message to confirm the actual type that was being used for the  `Real` type. Hadn't realized till now though that the default message mode appears to be a mode that's output in red (warning perhaps?). While this message isn't so helpful anymore, it's still a nice warm fuzzy for me but I'll change it to being a `STATUS` message which doesn't appear to output in red.

**Comment by louis-langholtz** [comment]
Added issue #19 for getting rid of the use of red for outputting purely informational messages.

**Comment by louis-langholtz** [comment]
By "four different build configurations" do you mean the four `CMAKE_CONFIGURATION_TYPES` of `Debug;Release;etc;etc`?

I wasn't sure about this so I looked up [`CMAKE_CONFIGURATION_TYPES`](https://cmake.org/cmake/help/v3.5/variable/CMAKE_CONFIGURATION_TYPES.html#variable:CMAKE_CONFIGURATION_TYPES) and then [`CMAKE_BUILD_TYPE`](https://cmake.org/cmake/help/v3.5/variable/CMAKE_BUILD_TYPE.html). Seems the `CMAKE_CONFIGURATION_TYPES` variable is set up by CMake to specify the build types available from CMake and is related to the `CMAKE_BUILD_TYPE` variable (which explained things better for me).

As far as I'm concerned, really only `Debug;Release` build configurations seem appealing where only `Release` should probably be what's built unless the user specifies more.

**Comment by NauticalMile64** [comment]
Ok, so this has been a somewhat nasty debugging process. Here is what I've found so far:

- Because the `glfw3` package is not CMake friendly there is no existing `FindGLEW3.cmake` script which helps to locate the library (unlike `GLEW`, for example). In this scenario, more primitive methods of searching for the library are used.
- One alternative to find these types of libraries is using the `PkgConfig` package, which also needs to be loaded on line 4 of `Testbed\CMakeLists.txt`. This package is slightly different in that it's looking for an executable file, and not a library. The user can direct CMake to the appropriate `FILEPATH` using the CMake variable `PKG_CONFIG_EXECUTABLE` or the environment variable `PKG_CONFIG` as eluded to [in the docs](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html). This gives access to the `pkg_search_module` function used on the next line. 

I've spent the rest of my time this evening trying to understand why, no matter how many times I try and set that variable and environment variable, it wasn't working. I fiddled around with the `CMake\share\cmake-3.9\Modules\FindPkgConfig.cmake` file until I was convinced that the `set` commands just weren't having any effect. After a few searches I found [this answer](https://stackoverflow.com/a/16497918/1542146), which hints at the fact that variables set in the cmake-gui are not overridden unless the `FORCE` flag is used. After the first time I tried the `FORCE` flag in a `set` call, the `PKG_CONFIG_EXECUTABLE` was set correctly, and it loaded the package properly. I believe the root has something to do with the `PKG_CONFIG_EXECUTABLE` variable behaving in a funny way. I've tried to reproduce the earlier errors, but even when I set `PKG_CONFIG_EXECUTABLE` to a nonsense string, it still finds the executable.

In any case, if a future user is struggling with this issue, I would recommend:

1. Create the `PKG_CONFIG_EXECUTABLE` in the gui and explicitly `Remove Entry` to try and unset the variable.
2. Add a set statement at the beginning of the `FindPkgConfig.cmake` script:
`set(PKG_CONFIG_EXECUTABLE "C:\path\to\pgk-config.exe" CACHE FILEPATH "pkg-config executable" FORCE)`

**Comment by NauticalMile64** [comment]
Even when this is done, the issue is still not resolved. The `pkg_search_module` fails, with the error report:

```
PLAYRHO_REALNUM_TYPE=float
Checking for one of the modules 'glfw3'
CMake Error at C:/Program Files (x86)/CMake/share/cmake-3.9/Modules/FindPkgConfig.cmake:640 (message):
  None of the required 'glfw3' found
Call Stack (most recent call first):
  Testbed/CMakeLists.txt:5 (pkg_search_module)


GLFW_FOUND=
GLFW_LIBRARY=GLFW_LIBRARY-NOTFOUND
GLFW_LIBRARY_DIRS=
Configuring incomplete, errors occurred!
```

This means the error is on my end, and likely something to do with `vcpkg` which is supposed to be CMake friendly, but it looks like I'll be posting an issue there, too.

**Comment by louis-langholtz** [comment]
Thanks for your efforts on this!

Does the information at http://www.glfw.org/docs/latest/build_guide.html#build_link_cmake_package help by any chance?

**Comment by NauticalMile64** [comment]
Unfortunately not 😓 . The `find_package` command fails as `glfw3` doesn't add a `FindGLFW3.cmake` file in order to locate it. For now I've commented out the `PkgConfig` lines and manually inserted the `GLFW_LIBRARY` and `GLFW_INCLUDE_DIRS` in the cmake-gui. It generates the VS project which builds a functional testbed in the `Debug` and `Release` configurations.

I'll leave this issue for now, and I'll see what the response is to the issue I submitted [here](https://github.com/Microsoft/vcpkg/issues/1570).

**Comment by louis-langholtz** [comment]
Please disregard the previously logged commit. I used the wrong issue in the commit message.

**Comment by Voskrese** [comment]
@NauticalMile64 
`set PKG_CONFIG_EXECUTABLE="C:\path\to\pkg-config.exe"`
`set PKG_CONFIG_PATH=E:\tools\vcpkg\installed\x86-windows\lib\pkgconfig;E:\tools\gstreamer\1.0\x86\lib\pkgconfig;E:\tools\vala-0.12.0\lib\pkgconfig;H:\Lib\gnome-deps\lib\pkgconfig` <<-- glfw3.pc

type glfw3.pc
`pkg-config --cflags glfw3`
-IE:/tools/vcpkg/installed/x86-windows/include

`pkg-config --cflags glfw3 glu`
-IE:/tools/vcpkg/installed/x86-windows/include -IE:/tools/vala-0.12.0/include

**Comment by NauticalMile64** [comment]
Resolving issue Microsoft/vcpkg#1570 has helped with this build, and I can load `glfw3` using `find_package(GLFW3 REQUIRED)`, and commenting out the `PKG_CONFIG` lines.

At this point I still need to add the library location in VS project (another `vcpkg` issue), and add the command line option `/NODEFAULTLIB:libcmt.lib` to quell an elevated warning error (libcmt conflicts with other libs). I don't remember having to deal with the libcmt error before. It might have been introduced when gtests was updated? In any case, the testbed builds after that.

------------

If I instead use the code as-is, CMake hangs and a prompt comes up saying

> `pkg-config` has stopped working.

@Voskrese I tried following your instructions, but I also get 'pkg-config has stopped working' when calling pkg-config from the command line.

**Comment by louis-langholtz** [comment]
@NauticalMile64  Is this resolved? Sorry, I'm not clear on that.

**Comment by NauticalMile64** [comment]
Well, `pkg-config` was absent from my machine, which caused the original issue. After discovering this, I thought the solution would just be to edit the docs to instruct users to install `pkg-config` from a dependable source and setup that CMake variable to point to the expected location, but it turns out `pkg-config` is not included with CMake, Visual Studio, or any other readily available source, and there's no clear tried and true way to install it properly in the first place. Making `pkg-config` a requirement raises the barrier of entry for Windows users, too much IMO.

So I'm thinking the best way to resolve this issue is to relieve Windows users of the need to install `pkg-config`, by trying other ways to direct CMake to the appropriate location of the `glew` library. After more learning about CMake and `find_package`, and finally [asking on SO](https://stackoverflow.com/q/45524666/1542146), I've realized that there's not going to be a simple solution. I'm working now to code something that will work for everyone, but it's going to be much lengthier than before.

I'm curious: how did this pass the appveyor test in the first place? Do they have working versions of `pkg-config`? Or does the appveyor not compile the testbed? Did you test it out on a local computer?

**Comment by louis-langholtz** [comment]
@NauticalMile64 If by "this" in your question you mean *PlayRho*, I've so far gotten  AppVeyor to build the library and just barely got the `HelloWorld` application for `Debug` and `Release` configurations for `Win32` and `x64` platforms building. These are built from a hand crafted [MSVS project file](https://github.com/louis-langholtz/PlayRho/blob/master/Build/vs2017/PlayRho.vcxproj) and AppVeyor itself is configured by the [`appveyor.yml`](https://github.com/louis-langholtz/PlayRho/blob/master/appveyor.yml) file.

Incidentally, CI on Travis-CI (for Linux) is further along than AppVeyor (for Windows) in this regard. 

I want AppVeyor to eventually build all the applications (`HelloWorld`, `Testbed`, `UnitTests`, and `Benchmark`) but I'm still working out how to test if appveyor's `HelloWorld` application build works correctly. AppVeyor doesn't seem to support console output from running a console application so I need to work out another way to check that `HelloWorld` outputs expected results. Once past that hurdle, the `UnitTests` application is next priority in my mind for AppVeyor. Then the `Testbed` application. And finally the `Benchmark` application.

I also want AppVeyor to build the MSVS project file from the `CMakeLists.txt` file and then build and test all the applications through that mechanism as well.

In summary, there's a lot for me to learn still about how to setup AppVeyor unless someone knows AppVeyor well already and helps out with it.

**Comment by louis-langholtz** [comment]
@NauticalMile64 Would it be easier to include binaries of some of these things within the `PlayRho` distribution? I've seen where some projects will do that like for needed libraries (SFML for instance does this). Seemingly that's a compromise of sorts but if it helps then perhaps it's worth considering. Alternatively, what if user's were instructed to download the Windows binaries into a well known folder/directory say within PlayRho where `CMakeLists.txt` knew to look for them when run on Windows.

**Comment by NauticalMile64** [comment]
Resolved by prs #30 and #54.
